Workiva/over_react
# FED-2139 Fix ProviderProps.value being optional/nullable, refactor internal impl

## Dependent PR

Dependent on https://github.com/Workiva/over_react/pull/875

---

## Motivation
When looking at context APIs in https://github.com/Workiva/over_react/pull/875, I realized that the value prop on Provider wasn't considered required (by runtime checks and static analysis), even though it should be. It was also `TValue?`, when it should have been `TValue`.



`ProviderProps.value` was previously implemented as a custom getter/setter, and required props can only be declared in generated props mixins.

This refactor moves away from manual implementations to relying on generated code, cleaning up certain aspects, but unfortunately the implementations are still pretty custom and messy.


## Changes
- Move declaration of `ProviderProps.value` from a custom getter/setter to a `late` required, potentially non-nullable (when as TValue is non-nullable) prop in a props mixin (required props can only be declared in generated props mixins)
- Refactor the custom implementations of ProviderProps/ConsumerProps to rely more on generated code. 
    - This cleans up certain aspects, but unfortunately the implementations are still pretty custom and messy due to the use of generics, and due to `ConsumerProps`'s `call` override. See code comments for more details.
- Add tests
   - that the generics are wired up properly
   - that the factories work properly when called with maps
   - for staticMeta (there was an existing TODO for those that became invalid with these changes)

#### Release Notes
  

## Review
_[See CONTRIBUTING.md][contributing-review-types] for more details on review types (+1 / QA +1 / +10) and code review process._

  

Please review: 

### QA Checklist
        - CI passes
        - Merge this branch and the branch in https://github.com/Workiva/over_react/pull/871 together, and verify in the analyzer playground that the analyzer lints about rendering a context provider without a `value` prop

## Merge Checklist
While we perform many automated checks before auto-merging, some manual checks are needed:


[contributing-review-types]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#review-types
[contributing-manual-testing]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#manual-testing-criteria
